In [2]:
df

order_id  order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214              1   
1       00018f77f2f0320c557190d7a144bdd3              1   
2       000229ec398224ef6ca0657da4fc703e              1   
3       00024acbcdf0a6daa1e931b038114c75              1   
4       00042b26cf59d7ce69dfabb4e55b4fd9              1   
...                                  ...            ...   
115604  fffc94f6ce00a00581880bf54a75a037              1   
115605  fffcd46ef2263f404302a634eb57f7eb              1   
115606  fffce4705a9662cd70adb13d4a31832d              1   
115607  fffe18544ffabc95dfada21779c9644f              1   
115608  fffe41c64501cc87c801fd61db3f6244              1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3       7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4       ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
...                                  ...                               ...   
115604  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
115605  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
115606  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
115607  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
115608  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

        shipping_limit_date   price  freight_value   product_category_name  \
0       2017-09-19 09:45:35   58.90          13.29              cool_stuff   
1       2017-05-03 11:05:13  239.90          19.93                pet_shop   
2       2018-01-18 14:48:30  199.00          17.87        moveis_decoracao   
3       2018-08-15 10:10:18   12.99          12.79              perfumaria   
4       2017-02-13 13:57:51  199.90          18.14      ferramentas_jardim   
...                     ...     ...            ...                     ...   
115604  2018-05-02 04:11:01  299.99          43.41   utilidades_domesticas   
115605  2018-07-20 04:31:48  350.00          36.53  informatica_acessorios   
115606  2017-10-30 17:14:25   99.90          16.95           esporte_lazer   
115607  2017-08-21 00:04:32   55.99           8.72  informatica_acessorios   
115608  2018-06-12 17:10:13   43.00          12.79         cama_mesa_banho   

                             customer_id order_status  ... payment_value  \
0       3ce436f183e68e07877b285a838db11a    delivered  ...         72.19   
1       f6dd3ec061db4e3987629fe6b26e5cce    delivered  ...        259.83   
2       6489ae5e4333f3693df5ad4372dab6d3    delivered  ...        216.87   
3       d4eb9395c8c0431ee92fce09860c5a06    delivered  ...         25.78   
4       58dbd0b2d70206bf40e62cd34e84d795    delivered  ...        218.04   
...                                  ...          ...  ...           ...   
115604  b51593916b4b8e0d6f66f2ae24f2673d    delivered  ...        343.40   
115605  84c5d4fbaf120aae381fad077416eaa0    delivered  ...        386.53   
115606  29309aa813182aaddc9b259e31b870e6    delivered  ...        116.85   
115607  b5e6afd5a41800fdf401e0272ca74655    delivered  ...         64.71   
115608  96d649da0cc4ff33bb408b199d4c7dcf    delivered  ...         55.79   

                      customer_unique_id customer_zip_code_prefix  \
0       871766c5855e863f6eccc05f988b23cb                    28013   
1       eb28e67c4c0b83846050ddfb8a35d051                    15775   
2       3818d81c6709e39d06b2738a8d3a2474                    35661   
3       af861d436cfc08b2c2ddefd0ba074622                    12952   
4       64b576fb70d441e8f1b2d7d446e483c5                    13226   
...                                  ...                      ...   
115604  0c9aeda10a71f369396d0c04dce13a64             

In [5]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix
import pandas as pd
df=pd.read_csv('data.csv')
df['price'] = pd.to_numeric(df['price'], errors='coerce')
df['freight_value'] = pd.to_numeric(df['freight_value'], errors='coerce')
df['sale'] = df['price'] + df['freight_value']
features = ['price', 'freight_value', 'review_score']
target = 'order_status'

# Chọn các cột có mặt số
X =df[features]
y = df[target]

# Chuyển đổi các cột mục tiêu thành số
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

# Chia dữ liệu thành tập huấn luyện và tập kiểm tra
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Tiền xử lý dữ liệu (chuẩn hóa)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)
# Khởi tạo mô hình
model = RandomForestClassifier(n_estimators=100, random_state=42)

# Huấn luyện mô hình
model.fit(X_train, y_train)

# Dự đoán trên tập kiểm tra
y_pred = model.predict(X_test)


# Chuyển đổi dự đoán về dạng văn bản (nếu cần)
y_pred_labels = label_encoder.inverse_transform(y_pred)

# Tạo DataFrame chứa các dự đoán
results = pd.DataFrame({
    'Actual': label_encoder.inverse_transform(y_test),
    'Predicted': y_pred_labels
})

# Hiển thị 20 dự đoán đầu tiên
print(results.head(20))
# Đánh giá mô hình
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

       Actual  Predicted
0   delivered  delivered
1   delivered  delivered
2   delivered  delivered
3   delivered  delivered
4   delivered  delivered
5   delivered  delivered
6   delivered  delivered
7   delivered  delivered
8   delivered  delivered
9   delivered  delivered
10  delivered  delivered
11  delivered  delivered
12  delivered  delivered
13  delivered  delivered
14  delivered  delivered
15  delivered  delivered
16  delivered  delivered
17  delivered  delivered
18  delivered  delivered
19  delivered  delivered
[[    1     0     0     0     0     0     0]
 [    1    38   110     2     2     1     0]
 [    1    36 33842    13    23    63     1]
 [    1     1    93    14     2     2     0]
 [    0     0    66     1    25     3     0]
 [    0     3   289     3     3    39     0]
 [    0     0     4     0     0     0     0]]
              precision    recall  f1-score   support

           0       0.25      1.00      0.40         1
           1       0.49      0.25      0.33       


Confusion Matrix (Ma trận nhầm lẫn) là một công cụ hữu ích trong việc đánh giá hiệu suất của mô hình phân loại. Nó giúp bạn hiểu cách mô hình của bạn hoạt động với từng lớp trong tập dữ liệu. Dưới đây là cách giải thích chi tiết về confusion matrix:

Cấu trúc của Confusion Matrix
Confusion matrix là một ma trận vuông, trong đó mỗi hàng đại diện cho lớp thực tế và mỗi cột đại diện cho lớp dự đoán của mô hình.

Ví dụ Ma Trận Nhầm Lẫn
Giả sử bạn có ma trận nhầm lẫn sau:

css
Sao chép mã
[[    1     0     0     0     0     0     0]
 [    1    38   110     2     2     1     0]
 [    1    36 33842    13    23    63     1]
 [    1     1    93    14     2     2     0]
 [    0     0    66     1    25     3     0]
 [    0     3   289     3     3    39     0]
 [    0     0     4     0     0     0     0]]
Mỗi hàng đại diện cho lớp thực tế và mỗi cột đại diện cho lớp dự đoán. Để dễ hiểu, chúng ta sẽ mô tả các phần của ma trận này:

Hàng đầu tiên: Lớp thực tế là lớp 0. Các giá trị trong hàng này cho biết số lượng mẫu thuộc lớp 0 được dự đoán thành các lớp khác:

[1, 0, 0, 0, 0, 0, 0]: Có 1 mẫu thực tế thuộc lớp 0 được dự đoán đúng là lớp 0. Không có mẫu nào bị dự đoán thành các lớp khác.
Hàng thứ hai: Lớp thực tế là lớp 1. Các giá trị trong hàng này cho biết số lượng mẫu thuộc lớp 1:

[1, 38, 110, 2, 2, 1, 0]: Trong số 154 mẫu thực tế thuộc lớp 1:
1 mẫu dự đoán là lớp 0.
38 mẫu dự đoán đúng là lớp 1.
110 mẫu dự đoán thành lớp 2.
2 mẫu dự đoán thành lớp 3, v.v.
Hàng thứ ba: Lớp thực tế là lớp 2. Tương tự:

[1, 36, 33842, 13, 23, 63, 1]: Trong số 33979 mẫu thực tế thuộc lớp 2:
1 mẫu dự đoán là lớp 0.
36 mẫu dự đoán là lớp 1.
33842 mẫu dự đoán đúng là lớp 2.
13 mẫu dự đoán thành lớp 3, v.v.
Cách Đọc Confusion Matrix
Chỉ số chính xác cho lớp i: ma trận[i][i] cho biết số lượng mẫu của lớp i mà mô hình đã dự đoán đúng.
Số lỗi cho lớp i: Tổng các giá trị trong hàng i trừ đi giá trị chính xác (sai sót được dự đoán thành các lớp khác).
Tổng số mẫu cho lớp i: Tổng các giá trị trong hàng i, cho biết tổng số mẫu thực tế thuộc lớp i.
Phân Tích
Precision cho lớp i: Tỷ lệ dự đoán đúng cho lớp i so với tổng số dự đoán là lớp i.
Recall cho lớp i: Tỷ lệ dự đoán đúng cho lớp i so với tổng số mẫu thực tế thuộc lớp i.
Confusion matrix giúp bạn phân tích hiệu suất mô hình phân loại của mình, đặc biệt là khi bạn có nhiều lớp và cần biết mô hình có xu hướng dự đoán sai cho lớp nào. Nếu có bất kỳ câu hỏi nào khác hoặc cần thêm thông tin, đừng ngần ngại hỏi!

Kết quả này là bảng đánh giá mô hình phân loại của bạn, được tạo ra bằng cách sử dụng các chỉ số như precision, recall, f1-score và support. Đây là cách bạn có thể hiểu các thông số trong bảng:

1. **Confusion Matrix**: 
   - Đoạn ma trận đầu tiên là confusion matrix, cho biết số lượng dự đoán đúng và sai cho từng lớp. Mỗi hàng tương ứng với lớp thực tế và mỗi cột tương ứng với lớp dự đoán.

2. **Classification Report**:
   - **Precision**: Tỷ lệ dự đoán chính xác trong số các dự đoán là lớp đó (ví dụ, precision của lớp 0 là 0.25 có nghĩa là 25% của các dự đoán lớp 0 là chính xác).
   - **Recall**: Tỷ lệ đúng các mẫu thuộc lớp đó được dự đoán đúng (ví dụ, recall của lớp 1 là 0.25 có nghĩa là 25% của các mẫu thực sự thuộc lớp 1 được dự đoán đúng).
   - **F1-Score**: Trung bình hài hòa của precision và recall (F1-score của lớp 0 là 0.40).
   - **Support**: Số lượng mẫu thực tế thuộc lớp đó trong tập kiểm tra.

3. **Accuracy**: Tỷ lệ chính xác tổng thể của mô hình (0.98 có nghĩa là 98% các dự đoán là chính xác).

4. **Macro Average**: Trung bình của các chỉ số (precision, recall, f1-score) trên tất cả các lớp, không cân nhắc đến tần suất của từng lớp.

5. **Weighted Average**: Trung bình các chỉ số trên tất cả các lớp, cân nhắc đến số lượng mẫu của từng lớp.

Bảng này giúp bạn hiểu rõ hơn về hiệu suất của mô hình trên từng lớp và tổng thể, giúp bạn đánh giá và cải thiện mô hình của mình. Nếu có vấn đề nào khác hoặc cần thêm chi tiết, hãy cho tôi biết!